Allow VFIO devices to be passed into container *as* VFIO devices.

dgibson opened the pull request:
This patches series implements a fix for #2938 

To work, it requires the corresponding agent changes in https://github.com/kata-containers/agent/pull/845

c3d commented on virtcontainers/qemu.go:
```diff
@@ -1275,6 +1275,59 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
+	}
+	addri := uint8(addrf)
+
+	if (addri & 0x7) != 0 {
```
What does this case correspond to? Why can't it be non-zero in our use case?

If we ever need it, it seems easy enough to make the function return `(uint8, uint8, err)` and add one extra `uint8` to its caller too. So I guess you probably had a good reason to consider it would always be zero.

c3d commented on virtcontainers/qemu.go:
```diff
@@ -1275,6 +1275,59 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
+	addri := uint8(addrf)
+
+	if (addri & 0x7) != 0 {
+		return 0xff, fmt.Errorf("Can't cope with non-zero function on %s", qomPath)
```
I would state "non-zero _PCI_ function"

c3d commented on virtcontainers/qemu.go:
```diff
@@ -1275,6 +1275,59 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
+		return 0xff, 0xff, fmt.Errorf("parent_bus QOM property of %s is %t not a string", qemuID, busq)
+	}
+
+	// Chop off the last QOM path component to go from the bus to
```
Might be valuable to explain the structure of the qom path you receive, or give a pointer to some doc that does.

c3d commented on virtcontainers/qemu.go:
```diff
@@ -1341,7 +1394,18 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) (err erro
+		if err != nil {
+			return err
+		}
+		// XXX: In some circumstances we already know some or
```
As written, the comment confuses me more than it enlightens me, because it talks about "some circumstances" that I have no real clue about. I can _guess_ that you talk about the path above where you had a `bridge` variable, for example, but even so, that declaration is local to the scope, so you can't use it here.

Not that it would be a biggie to change, but the inconsistency between the comment and the scoping muddies the waters as what you mean with "we know" - Yes, we computed it above, but we no longer have it.

Also, I think that going back to qom allows you to get the info from qemu direcly, whereas otherwise, you would just infer it.

dgibson replied:
The variable scope isn't really the issue - that's easily altered.  The two cases here are plugging onto a bridge and plugging into a root port.

For a bridge, we actually allocate a slot ourselves (addDeviceToBridge()) and tell qemu to use it, so we already know that.  The bridge structure itself also has an address field.... but it's not actually filled in properly, at least in our case. Sigh.

For a root port we don't set an address for the device , but it will  always be zero, because, well, it's a port, there's only one place to go.  However, the root port itself has its address allocated by qemu so Kata doesn't know it until we ask qemu.  For extra inconsistency the handling of the root ports is more or less open-coded, not using the sort-of-general-but-not-really bridge handling code.  Sigh, again.

Hence the decision to just query it out for all cases.  I very much want to clean this up, but that's a later project for Kata2.

amshinde replied:
> The bridge structure itself also has an address field.... but it's not actually filled in properly, at least in our case. Sigh.
Can you point to where this is happening?

Also, same as @c3d, I was confused after reading the comment here. Can you add the above explanation to the comments here. As far as hotplug goes, there are 3 scenarios we deal with today - pci root bus, pci/pcie bridge and pcie root ports. I want to make sure all those scenarioes are taken into account here.

c3d commented on cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -466,3 +466,17 @@ sandbox_cgroup_only=@DEFSANDBOXCGROUPONLY@
+# different name from the device on the host, since the VM has its own
+# IOMMU/VFIO group numbering.
+#
+# When disabled, the VM will use its native drivers for the VFIO
```
I wrote elsewhere that I was expecting this to be per-device and not per-containers.

I see based on this comment that you really mean that as a sort of "backward compatibility" fallback. In that case, it makes more sense. That being said, I don't really understand how the container would use the VFIO device without having a proper name for it.

Without that flag, is there any way to get something sensible in the guest at all?

dgibson replied:
Right to use a VFIO device in the old behaviour the container payload would have to be expecting it explicitly.  It might also need other hacks like mapping the VM's /dev into the container.

amshinde replied:
How about `use_vfio_in_guest` or `use_vfio_driver_in_guest` since we tend to use a verb based config for turning a particular feature on/off ?

c3d commented on cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -466,3 +466,17 @@ sandbox_cgroup_only=@DEFSANDBOXCGROUPONLY@
+# device.This behaviour is inherently Kata specific, and will require
+# a container image which expects it.
+# (default: false)
+#vfio_in_guest = true
```
Having received that feedback from @fidencio not very long ago, I would recommend you add a variable to the `Makefile` and add something that looks like `@DEFVIRTIOFSINGUEST@` in the `.in` files. That way, the various distros can more easily define their own default values.

Something like:
```
# default if not set: false
# Your distribution recommends: @DEFVIRTIOFSINGUEST@
vfio_in_guest = @DEFVIRTIOFSINGUEST@
```

c3d commented on cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -466,3 +466,17 @@ sandbox_cgroup_only=@DEFSANDBOXCGROUPONLY@
 # (default: [])
 experimental=@DEFAULTEXPFEATURES@
+
+# Use VFIO driver for VFIO devices in the VM, as well as the host.
```
Based on the description, I would call the variable `use_guest_driver_for_vfio_devices` or something like that. You are not enabling VFIO in the guest, you are changing how the driver is bound.

dgibson replied:
`use_guest_driver_for_vfio_devices` is ambiguous, because it's not clear if you're using the guest's VFIO driver or the guest's "normal" driver for the device, which is precisely the issue.  Originally I wanted to call this `use_guest_native_driver_for_vfio` with the sense reversed (so true for the old behaviour, false for the new).  But for irritating reasons I couldn't figure out a sane way to add a boolean config option that defaulted to true.

c3d replied:
What about `unbind_guest_vfio_drivers`?

dgibson replied:
That's even worse: we don't want to unbind the vfio drivers, we want to unbind the not-vfio drivers for the vfio devices.

c3d reviewed (changes requested):
Overall, looks quite good now, and I like the overall approach.

However, I think that there is one bug worth fixing, a problem with scoping in a `defer` in the commit where you did some refactoring.

I would also like the variables in the makefiles to be added, for consistency with the rest of the settings.

The rest is more cosmetic and details.

dgibson commented on virtcontainers/kata_agent.go:
```diff
@@ -1148,17 +1149,25 @@ func (k *kataAgent) constraintGRPCSpec(grpcSpec *grpc.Spec, passSeccomp bool) {
+	// leave them in, so the agent can use the information here
+	// and elsewhere to properly bind drivers and create device
+	// node in the container.
+	if !vfioInGuest {
```
> Why is `vfioInGuest` a per-container setting and not per vfio device?

Well, it's not even really a per-container setting.  It's pretty much a global Kata setting that's just copied into the container.

Per-device would be nice, but I can't really see a way to do that.  The only naturally per-device information we have is the device info from the OCI spec and that doesn't really have a way to add options.  Even if it did, the default would really have to be the new behaviour.

> I think that ultimately, you want two levels:
> 
>     * A system-wide default value, for the compatibility reasons you explained
> 
>     * A per-device value that allows you to have a configuration where one device would use the guest driver and the other would be rebound to VFIO.

Well, sure, but we have to figure out how to do that.  Really the only way to do this is to invoke the *old* behaviour with a new syntax in the OCI spec that's explicitly a Kata extension, while leaving the existing form to invoke the new behaviour.

> Even if that is true, that can obviously wait a next PR.

Indeed.

c3d replied:
OK. Marking conversation as resolved then.

dgibson commented on virtcontainers/kata_agent.go:
```diff
@@ -1253,6 +1262,44 @@ func (k *kataAgent) appendVhostUserBlkDevice(dev ContainerDevice, c *Container)
+	// Each option is formatted as "DDDD:BB:SS.F=NN/MM"
+	// DDDD:BB:SS.F is the device's PCI address on the *host*.  NN
+	// is PCI slot number of the device's bridge on the guest, and
+	// MM is the slot number of the device itself on the guest.
```
Thank you, though that's not a very high bar :/.

dgibson commented on virtcontainers/qemu.go:
```diff
@@ -1312,33 +1312,36 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) (err erro
-			return q.qmpMonitorCh.qmp.ExecutePCIVFIOMediatedDeviceAdd(q.qmpMonitorCh.ctx, devID, device.SysfsDev, addr, bridge.ID, romFile)
-		default:
-			return fmt.Errorf("Incorrect VFIO device type found")
+			defer func() {
```
Ugh.  Right.  I noticed already that the way := interacts with variable shadowing can be rather surprising.  So I'm not surprised I made an mistake here.

dgibson commented:
New push is a significant rework based on #2981 .
/test-ubuntu

c3d commented:
@dgibson I see you pushed new updates, but they don't yet address all my remarks. I believe you are not done yet. Asking because GitHub shows that you requested my review. I suspect GitHub automatically did that when you pushed the new version. Let me know when you want me to review again. Thanks.

dgibson commented:
/retest-ubuntu

dgibson commented:
> @dgibson I see you pushed new updates, but they don't yet address all my remarks. I believe you are not done yet. Asking because GitHub shows that you requested my review. I suspect GitHub automatically did that when you pushed the new version. Let me know when you want me to review again. Thanks.

Well, if you could first look at #2981 and https://github.com/kata-containers/agent/pull/850 first, which are for different (prerequisite) issues but contain a good chunk of the code I was working on.  I thought I'd addressed your remarks for that part of the code, but the rework is substantial enough that it's possible I've missed the continued applicability of something.

amshinde commented on virtcontainers/container.go:
```diff
@@ -886,38 +886,9 @@ func (c *Container) create() (err error) {
 		}
 	}
 
-	var (
```
I need to go back to the discussion on the  agent PR to see its progress, but adding a note here that the agent PR was blocked due to the inherent delay introduced by the kernel while hotplugging pcie devices.

amshinde commented on virtcontainers/kata_agent.go:
```diff
@@ -1149,17 +1150,25 @@ func (k *kataAgent) constraintGRPCSpec(grpcSpec *grpc.Spec, passSeccomp bool) {
+		var linuxDevices []grpc.LinuxDevice
+		for _, dev := range grpcSpec.Linux.Devices {
+			if dev.Type == "c" && strings.HasPrefix(dev.Path, vfioPath) {
+				k.Logger().WithField("vfio-dev", dev.Path).Debug("removing vfio device from grpcSpec")
```
Why are you removing the vfio devices here? I though you would simply replace them in the agent.

Seeing your agent PR after the fact, will need to look at that to see how you are handling this.

dgibson replied:
There was already code to remove the devices here, I'm just making it conditional on the configuration.

dgibson commented:
@amshinde regarding the float vs. int in the qom get code.  The problem is that JSON doesn't actually distinguish floats from integers, they're both just "number".  So when decoding to an interface{} we always get a float, even though the data in there should be an integer.  It sort of depends on your interpretation, but arguably json is only safe to convey integer values which can be losslessly encoded in a 64-bit float (which includes the whole range of a u32 or i32, but not of an i64 or u64).

jodh-intel commented on virtcontainers/qemu.go:
```diff
@@ -1275,6 +1275,59 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
+	// XXX going via float makes no real sense, but that's how
+	// JSON works, and we should get away with it for the small
+	// values we have here
+	if !ok {
```
@dgibson - I've just looked at the qemu code and it *appears* that the returned address is actually a JSON string, not a JSON number (aka float). Please could you check that?

dgibson replied:
I'm pretty sure it's a number - you may be looking at the `legacy_addr` property rather than `addr`.  I don't think the things I'm trying with this could possibly work if that were not the case.  But I will double check.

fidencio commented:
Folks, this PR has been stale and I'm closing it as we're walking towards the **final** 1.13.0 release of the kata-containers-1.x.
From now on, all the development will happen as part of https://github.com/kata-containers/kata-containers/ and fixes should be provided to that repo.

Thanks a lot for your contribution, we're sorry that for some reason it got stale, and we hope to see you around for new adventures in the kata-containers-2.x land.
